content/blog/2024/09/usr-merge/index.md

```diff
@@ -0,0 +1,90 @@
+   
+   Because of the time the script could take to run, a message will be shown on screen so the user knows that something is happening. The script is robust enough to not destory your system if a reboot occurs while it is running, but doing a hard reboot does mean that the script has to start over on the next boot.
+   
+   Lastly, a marker file is written that indicates that the system is now Usr Merged.
+3) eopkg will use the second marker file to know that it is safe to switch to a new package repository that is incompatible with non-Usr-Merged systems.
+
+   The switch will be activated when the phased rollout is complete.
+
+Once the migration has rolled out to everyone (more on this later), we can fork our package repository and do an Epoch bump, where we can safely release package updates, knowing that they'll only be going out to Usr Merged systems. While an Epoch bump isn't strictly necessary for the merging process itself, it *is* necessary to safely remove our pre-Usr-Merge compatability symlinks from various packages. Removing the symlinks on a system that has not been merged will likely break the system, as we saw last time, because update order is non-deterministic.
+
+Migration to a new package repository has actually been done before on Solus. Most recently, it was done when the [repositories were moved](https://getsol.us/2018/09/11/package-repo-migration-now-available/) to the infrastructure at the Rochester Institute of Technology.
+
+### Staged rollout
+
+We're aiming for a staged rollout, where everyone gets the necessary updates, but the script will only run on some systems, leaving everyone else operating as before. After our last attempt, we want to ensure that the process is as robust as possible, and if there is another problem that we have not forseen, it will only affect a limited number of users, and the rollout can be stopped with minimal disruption. Assuming that no issues pop up, each weekly sync after the rollout begins will see more machines going through the Usr Merge process, until no one is being held back.
+
+A staged rollout can be achieved by taking the systemd machine ID of a computer, deriving a number from it between 0 and 255, and checking if the derived number is below a certain threshold. If it is, then the merging script continue run; otherwise, the script will exit, and the system will not be Usr Merged. The magic number can be set in a file that gets shipped with the script, enabling us to update it for each rollout stage with the weekly sync.
+
+The entire process goes like this:
+
+Stage 1: Allow opt-in to Usr Merge via a configuration file.
+Stage 2: Allow opt-in, and merge *x* amount of computers.
+Stage 3..*n*: Each sync, more systems are allowed to go through the process.
+Stage *n+1*: All users have been merged; do the Epoch bump. Users start getting updates from the new repository, and everyone is on a Usr Merged system! :tada:
+
```

Comment:
Ah, just saw that the above list doesn't get rendered as a list, but as a paragraph (contrary to my previous "stuff looks fine visually comment). Also the :tada: emoji doesn't get rendered.